
## Main [Langchain Components](https://python.langchain.com/docs/integrations/components) for interacting with LLM

- 🗃️ LLMs
- 🗃️ Chat models
- 🗃️ Embedding models
- 🗃️ Document loaders
- 🗃️ Document transformers
- 🗃️ Vector stores
- 🗃️ Retrievers
- 🗃️ Tools
- 🗃️ Toolkits
- 🗃️ Memory
- 🗃️ Callbacks
- 🗃️ Chat loaders
- 🗃️ Adapters
- 🗃️ Stores

#### API Keys, Websites, and Models for Cohere, Google AI, and Hugging Face


|            | API Keys    | Websites                        | Models                            |
|------------|-------------|---------------------------------|-----------------------------------|
| Cohere     | cohere_api_key | [Cohere Website](https://cohere.ai/) | Cohere Models                    |
| Google AI  | google_api_key | [Google AI Website](https://makersuite.google.com/app/apikey) | Google AI Models                 |
| Hugging Face | huggingface_api_key | [Hugging Face Website](https://huggingface.co/) | Hugging Face Models           |

**Note: After obtaining API keys, securely store them in Google Colab. For instructions on how to do this, refer to this  [Guide](https://drlee.io/how-to-use-secrets-in-google-colab-for-api-key-protection-a-guide-for-openai-huggingface-and-c1ec9e1277e0)**

In [22]:
## Get API Keys from Colab User data
from google.colab import userdata
import os
os.environ["LANGCHAIN_TRACING_V2"] = "false"

cohere_api_key= userdata.get("COHERE_API_KEY")
GOOGLE_API_KEY= userdata.get("GOOGLE_API_KEY")
hugging_face_api_key = userdata.get("HUGGINGFACEHUB_API_KEY")

LANGCHAIN_API_KEY=userdata.get("LANGCHAIN_API_KEY")


## 1. [Google Open Source LLM](https://ai.google.dev/tutorials/python_quickstart) ==> Gemini

In [8]:
import pathlib
import textwrap

import google.generativeai as genai

from IPython.display import display
from IPython.display import Markdown


def to_markdown(text):
  text = text.replace('•', '  *')
  return Markdown(textwrap.indent(text, '> ', predicate=lambda _: True))

In [9]:
## Configure API key
import getpass
import os
import google.generativeai as genai

## To Setup Google API key
# os.environ['GOOGLE_API_KEY'] = GOOGLE_API_KEY

# if "GOOGLE_API_KEY" not in os.environ:
#    os.environ["GOOGLE_API_KEY"] = getpass.getpass("Provide your Google API Key")

## We can directly get the API Key using like this
genai.configure(api_key=GOOGLE_API_KEY)



In [10]:
## List of Google Models
for m in genai.list_models():
    if 'generateContent' in m.supported_generation_methods:
        print(m.name)

models/gemini-1.0-pro
models/gemini-1.0-pro-001
models/gemini-1.0-pro-latest
models/gemini-1.0-pro-vision-latest
models/gemini-pro
models/gemini-pro-vision


In [ ]:
model = genai.GenerativeModel('gemini-pro')

In [ ]:
%%time
response = model.generate_content("What is the meaning of life?")

"""
response = model.generate_content(
    "What is the meaning of life?",
    generation_config=genai.types.GenerationConfig(
        # Only one candidate for now.
        candidate_count=1,
        stop_sequences=['x'],
        max_output_tokens=20,
        temperature=1.0,)
)
"""

CPU times: user 97.8 ms, sys: 11.4 ms, total: 109 ms
Wall time: 7.83 s


'\nresponse = model.generate_content(\n    "What is the meaning of life?",\n    generation_config=genai.types.GenerationConfig(\n        # Only one candidate for now.\n        candidate_count=1,\n        stop_sequences=[\'x\'],\n        max_output_tokens=20,\n        temperature=1.0,)\n)\n'

In [ ]:
## To get the text
response.text

"The meaning of life is a philosophical question that has been pondered by humans for centuries. There is no one definitive answer to this question, as it is a deeply personal one that can vary depending on individual beliefs, values, and experiences. Some common themes that emerge in discussions about the meaning of life include:\n\n**Purpose and Meaningful Activity:** Many people find meaning in their lives through their work, hobbies, volunteering, or other activities that bring them a sense of purpose and fulfillment.\n**Relationships and Connections:** Human relationships are a central aspect of life for many people, and finding love, companionship, and community can bring significant meaning.\n**Personal Growth and Development:** Striving for self-improvement, learning new things, and expanding one's horizons can give life meaning and a sense of accomplishment.\n**Contribution to Society:** Making a positive impact on society, whether through acts of kindness, social activism, or

In [ ]:
## To get the text in Markdown format
to_markdown(response.text)

> The meaning of life is a philosophical question that has been pondered by humans for centuries. There is no one definitive answer to this question, as it is a deeply personal one that can vary depending on individual beliefs, values, and experiences. Some common themes that emerge in discussions about the meaning of life include:
> 
> **Purpose and Meaningful Activity:** Many people find meaning in their lives through their work, hobbies, volunteering, or other activities that bring them a sense of purpose and fulfillment.
> **Relationships and Connections:** Human relationships are a central aspect of life for many people, and finding love, companionship, and community can bring significant meaning.
> **Personal Growth and Development:** Striving for self-improvement, learning new things, and expanding one's horizons can give life meaning and a sense of accomplishment.
> **Contribution to Society:** Making a positive impact on society, whether through acts of kindness, social activism, or other forms of service, can give people a sense of purpose and meaning.
> **Living in Harmony with Nature:** Connecting with the natural world, appreciating its beauty, and striving to protect our environment can provide meaning for some people.
> **Finding Spiritual Fulfillment:** For many, spirituality or religion plays a central role in their sense of meaning and purpose.
> **Experiencing the Present Moment:** Appreciating the present moment and living in the here and now can help people find meaning and joy in life.
> **Creating a Legacy:** Leaving a lasting impact on the world or on future generations through one's work, relationships, or other contributions can provide a sense of meaning and purpose.
> 
> Ultimately, the meaning of life is a question that each individual must answer for themselves based on their own values, beliefs, and experiences. There is no right or wrong answer, and what gives one person meaning may not be meaningful to another.

In [ ]:
response.prompt_feedback

### Streaming in LLMs

Streaming in LLMs refers to the process of delivering response token by token, enabling immediate presentation of results to users without waiting for the entire generation to complete.

Large Language Models (LLMs) incorporate a streaming response mechanism, allowing processing to begin as soon as initial fragments of the response are available, rather than waiting for the entire response. This capability is particularly useful for showcasing evolving responses to users in real-time during the generation process, or for facilitating concurrent processing alongside response generation.

In [ ]:
%%time
response = model.generate_content("What is the meaning of life?", stream=True)

CPU times: user 94.2 ms, sys: 10.8 ms, total: 105 ms
Wall time: 7.42 s


In [ ]:
for chunk in response:
    print(chunk.text)
    print("_"*80)

The meaning of life is a profoundly personal and subjective question that has been pondered by
________________________________________________________________________________
 philosophers, theologians, and individuals throughout history. While there is no one universally accepted answer, some common themes that emerge include:

* **Purpose:** Finding a
________________________________________________________________________________
 higher purpose or goal that gives life meaning and direction. This could be related to a career, creative endeavors, relationships, or making a positive impact on the world.

* **Growth and Fulfillment:** Striving to develop oneself, learn, and experience new things. This involves pursuing personal growth, intellectual stimulation, and emotional
________________________________________________________________________________
 and spiritual fulfillment.

* **Contribution:** Making a meaningful contribution to society or others. This can involve volunteerin

## Chat Conversations with Gemini

In [ ]:
model = genai.GenerativeModel('gemini-pro')
chat = model.start_chat(history=[])
chat

In [ ]:
response = chat.send_message("In one sentence, explain how a computer works to a young child.")
to_markdown(response.text)

> A computer is like a magic box that can help you do all kinds of things by remembering information and following your instructions.

In [ ]:
chat.history

[parts {
   text: "In one sentence, explain how a computer works to a young child."
 }
 role: "user",
 parts {
   text: "A computer is like a magic box that can help you do all kinds of things by remembering information and following your instructions."
 }
 role: "model"]

In [ ]:
response = chat.send_message("Okay, how about a more detailed explanation to a high schooler?", stream=True)

for chunk in response:
    print(chunk.text)
    print("_"*80)

A computer is an electronic device that can be programmed to carry out a set of
________________________________________________________________________________
 instructions. It has four main components: the central processing unit (CPU), the memory, the storage devices, and the input/output (I/O)
________________________________________________________________________________
 devices. The CPU is the "brain" of the computer and controls all of its operations. The memory stores the instructions and data that the CPU is currently working on. The storage devices store data that is not currently being used by the CPU. The I/O devices allow the computer to communicate with the outside world,
________________________________________________________________________________
 such as through a keyboard, mouse, or printer.

When you give a computer a command, the CPU fetches the instructions for that command from the memory. It then executes those instructions, one by one. The CPU can perform 

In [ ]:
## Total Conversation History
chat._history

[parts {
   text: "In one sentence, explain how a computer works to a young child."
 }
 role: "user",
 parts {
   text: "A computer is like a magic box that can help you do all kinds of things by remembering information and following your instructions."
 }
 role: "model"]

## Embeddings

| Task Type           | Description                                                                                           |
|---------------------|-------------------------------------------------------------------------------------------------------|
| RETRIEVAL_QUERY     | Specifies the given text is a query in a search/retrieval setting.                                    |
| RETRIEVAL_DOCUMENT  | Specifies the given text is a document in a search/retrieval setting. Using this task type requires a title. |
| SEMANTIC_SIMILARITY | Specifies the given text will be used for Semantic Textual Similarity (STS).                           |
| CLASSIFICATION      | Specifies that the embeddings will be used for classification.                                        |
| CLUSTERING          | Specifies that the embeddings will be used for clustering.                                             |

In [ ]:
## For Single Senetece embedding
result = genai.embed_content(
    model="models/embedding-001",
    content="What is the meaning of life?",
    task_type="retrieval_document",
    title="Embedding of single string")

# 1 input > 1 vector output
print(str(result['embedding'])[:50], '... TRIMMED]')

[-0.003216741, -0.013358698, -0.017649598, -0.0091 ... TRIMMED]


In [ ]:
## For Multiple sentence Embedding
result = genai.embed_content(
    model="models/embedding-001",
    content=[
      'What is the meaning of life?',
      'How much wood would a woodchuck chuck?',
      'How does the brain work?'],
    task_type="retrieval_document",
    title="Embedding of list of strings")

# A list of inputs > A list of vectors output
for v in result['embedding']:
  print(str(v)[:50], '... TRIMMED ...')

[0.0040260437, 0.004124458, -0.014209415, -0.00183 ... TRIMMED ...
[-0.004049845, -0.0075574904, -0.0073463684, -0.03 ... TRIMMED ...
[0.025310587, -0.0080734305, -0.029902633, 0.01160 ... TRIMMED ...


In [ ]:
## lets check the Dimensions of Embeddings
len(result['embedding'][0])

768

**Note: Default size of embeddings in transformers typically depends on the specific transformer model architecture being used. However, a common default size for embeddings in many transformer-based models is 768 dimensions for models like BERT, RoBERTa, and DistilBERT.**

## [Acessing Google AI chat models using Langchain](https://python.langchain.com/docs/integrations/chat/google_generative_ai)

In [ ]:
import locale
def getpreferredencoding(do_setlocale = True):
    return "UTF-8"
locale.getpreferredencoding = getpreferredencoding

%pip install --upgrade --quiet  langchain-google-genai pillow -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 18.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 252.4/252.4 kB 20.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.2/64.2 kB 7.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.5/138.5 kB 15.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
imageio 2.31.6 requires pillow<10.1.0,>=8.3.2, but you have pillow 10.2.0 which is incompatible.


In [ ]:
from langchain_google_genai import GoogleGenerativeAI

llm = GoogleGenerativeAI(model="gemini-pro", google_api_key=GOOGLE_API_KEY)
print(
    llm.invoke(
        "What are some of the pros and cons of Python as a programming language?"
    )
)

**Pros:**

* **Beginner-friendly:** Python's simple syntax and readable code make it easy for beginners to learn and understand.
* **Versatile:** Python is a general-purpose language suitable for various applications, including web development, data science, machine learning, and scripting.
* **Extensive library support:** Python has a vast collection of libraries and frameworks that provide out-of-the-box functionality for various tasks.
* **Cross-platform:** Python programs can run on different operating systems (Windows, macOS, Linux) without the need for recompilation.
* **Dynamically typed:** Python is a dynamically typed language, which allows for greater flexibility and rapid development.
* **Interpreted:** Python is an interpreted language, making it easier to debug and iterate on code.

**Cons:**

* **Performance:** Compared to compiled languages, Python can be slower due to its interpreted nature.
* **Memory usage:** Python programs tend to consume more memory than compiled l

In [ ]:
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-pro",google_api_key=GOOGLE_API_KEY)
result = llm.invoke("I want you to act as a poet who is sitting near the sea in the evening, listening to the rhythmic waves crashing against the shore. I would like you to capture the essence of the sea in your words and create a beautiful poem that evokes the emotions and sensations of being by the sea. ")
print(result.content)

**Ode to the Sea**

In twilight's embrace, where whispers dance,
I sit by the shore, lost in a trance.
The waves' rhythmic crash, a symphony's call,
Inviting me to surrender, to heed their thrall.

A salty breeze caresses my face,
Whispering secrets of time and space.
The sun, a golden orb, sinks below the line,
Casting hues of amber on the ocean's brine.

The sea, a canvas of blues and greens,
Mirrors the vastness of my dreams.
Its depths conceal mysteries untold,
Stories of ships and sailors bold.

Its surface shimmers, a diamond's gleam,
Reflecting the moon's ethereal beam.
The waves, like dancers, sway and glide,
Their movements graceful, their rhythm a guide.

I feel the sea's embrace, its gentle sway,
Washing away my cares, my troubles at bay.
Its lullaby soothes my weary soul,
Restoring balance, making me whole.

Oh, sea, your beauty knows no bounds,
A source of solace, where peace abounds.
Your waves, your whispers, a symphony divine,
A testament to nature's power and design.



## 2. [COHERE Model](https://cohere.com/)

### [Guides and Concepts by COHERE platform about Generative AI](https://docs.cohere.com/docs/the-cohere-platform)
### [Different Models available in Cohere](https://dashboard.cohere.com/playground/chat)

## [Using Cohere Models with COHERE Platform](https://docs.cohere.com/reference/about)

In [ ]:
pip install cohere -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 1.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 29.5 MB/s eta 0:00:00


In [ ]:
import cohere
# set up the Cohere client.
co = cohere.Client(cohere_api_key)
response = co.chat(
  chat_history=[
    {"role": "USER", "message": "Who discovered gravity?"},
    {"role": "CHATBOT", "message": "The man who is widely credited with discovering gravity is Sir Isaac Newton"}
  ],
  message="What year was he born?",
  # perform web search before answering the question. You can also use your own custom connector.
  connectors=[{"id": "web-search"}]
)
print(response)

cohere.Chat {
	id: e84b4b67-c7eb-4434-b85a-e9bdaa65791a
	response_id: e84b4b67-c7eb-4434-b85a-e9bdaa65791a
	generation_id: 6d2a341b-fa2f-4156-b138-3c686306a107
	message: What year was he born?
	text: Isaac Newton was born on January 4 1643 according to the Gregorian calendar. This was recorded as December 25 1642 according to the Julian calendar, which was used in England at the time. This means his birth date is recorded as both December 25 and January 4.
	conversation_id: None
	prompt: None
	chat_history: None
	preamble: None
	client: <cohere.client.Client object at 0x780ac35eefb0>
	token_count: {'prompt_tokens': 1478, 'response_tokens': 54, 'total_tokens': 1532, 'billed_tokens': 77}
	meta: {'api_version': {'version': '1'}, 'billed_units': {'input_tokens': 23, 'output_tokens': 54}}
	is_search_required: None
	citations: [{'start': 25, 'end': 34, 'text': 'January 4', 'document_ids': ['web-search_7:3', 'web-search_4:1', 'web-search_1:69', 'web-search_4:30']}, {'start': 35, 'end': 39, 't

In [ ]:
to_markdown(response.text)

> Isaac Newton was born on January 4 1643 according to the Gregorian calendar. This was recorded as December 25 1642 according to the Julian calendar, which was used in England at the time. This means his birth date is recorded as both December 25 and January 4.

In [ ]:
# Create another prompt
message = "Write about Quntum Computing and how can we do qunatum computing with machine learning Models and what should be our assumptions ."

response = co.chat(
    model='command-nightly',
    message=message,
)
print(response.text)

Quantum computing is an emerging technology that harnesses the laws of quantum mechanics to solve problems too complex for classical computers. Quantum computers use quantum bits, or qubits, to encode information and process vast amounts of data in parallel, enabling exponentially faster computation for certain tasks.

Quantum machine learning (QML) is an interdisciplinary field combining quantum computing with machine learning techniques to address complex challenges. By leveraging quantum phenomena, QML aims to solve challenging machine learning problems, such as enhancing data classification, clustering, and regression techniques.

To perform quantum computing with machine learning models, the following steps can be taken:

1. Quantum Data Encoding: First, classical machine learning data must be converted and encoded into a quantum representation. This process involves mapping the classical data onto qubits, utilizing techniques such as amplitude encoding or angle encoding to repres

In [ ]:
## Get the response in a markdown format
to_markdown(response.text)

> Quantum computing is an emerging technology that harnesses the laws of quantum mechanics to solve problems too complex for classical computers. Quantum computers use quantum bits, or qubits, to encode information and process vast amounts of data in parallel, enabling exponentially faster computation for certain tasks.
> 
> Quantum machine learning (QML) is an interdisciplinary field combining quantum computing with machine learning techniques to address complex challenges. By leveraging quantum phenomena, QML aims to solve challenging machine learning problems, such as enhancing data classification, clustering, and regression techniques.
> 
> To perform quantum computing with machine learning models, the following steps can be taken:
> 
> 1. Quantum Data Encoding: First, classical machine learning data must be converted and encoded into a quantum representation. This process involves mapping the classical data onto qubits, utilizing techniques such as amplitude encoding or angle encoding to represent the information quantumly.
> 
> 2. Quantum Model Training: Next, the encoded data is used to train quantum machine learning models. Quantum algorithms, such as Grover's algorithm or quantum support vector machines, are utilized to train the model on the quantum data. This step aims to optimize the model's performance and accuracy on the quantum computer.
> 
> 3. Quantum Model Inference: After training, the quantum model can be utilized for inference tasks. The trained model is deployed on a quantum computer, where it can process quantum data and generate predictions or classifications based on the learned patterns. The results are then measured and decoded back into classical information.
> 
> 4. Comparison and Evaluation: Finally, the performance of the quantum machine learning model is compared to classical machine learning models. This step involves evaluating the accuracy, speed, and overall effectiveness of the quantum model in relation to classical counterparts. This comparison helps to validate the advantages and potential applications of quantum computing in machine learning tasks.
> 
> For successful quantum computing with machine learning, certain assumptions can be made:
> 
> 1. Speed and Parallelism: Quantum computers offer exponential speed advantages for specific problems, assuming efficient algorithms are utilized to harness parallel processing capabilities.
> 
> 2. Enhanced Data Processing: Quantum computing can provide improved data processing and analysis, particularly for complex datasets with intricate patterns and high-dimensional structures.
> 
> 3. Quantum Algorithm Efficiency: The effectiveness of QML relies on the development and optimization of quantum algorithms to efficiently solve specific machine learning problems. Research in this area aims to create robust and efficient quantum algorithms for various tasks.
> 
> 4. Scalable Qubits: The success of quantum computing depends on the availability of scalable and reliable qubits. Advancements in qubit technology, such as error correction and fault-tolerant designs, are crucial for building practical quantum computers.
> 
> 5. Domain-Specific Applications: Initial applications of QML are likely to be domain-specific, focusing on areas such as optimization, data analysis, and simulation, where the advantages of quantum computing can be leveraged to provide significant performance gains.
> 
> In summary, combining quantum computing with machine learning holds great potential for solving complex problems. While challenges remain in algorithm development, qubit scalability, and error correction, the field of QML is actively working to overcome these obstacles, driving the advancement of quantum-enhanced machine learning applications. As research and technology continue to evolve, QML may revolutionize various domains and provide powerful tools for data analysis and problem-solving.

## [ Using Cohere Models with Langchain Framework](https://python.langchain.com/docs/integrations/providers/cohere)

In [ ]:
pip install langchain-community -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 7.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 4.4 MB/s eta 0:00:00


In [ ]:
## Chat
from langchain_community.chat_models import ChatCohere
from langchain_core.messages import HumanMessage
chat = ChatCohere(cohere_api_key=cohere_api_key)
messages = [HumanMessage(content="knock knock")]
print(chat(messages))

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:117: LangChainDeprecationWarning: The function `__call__` was deprecated in LangChain 0.1.7 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(


content="Who's there?"


In [ ]:
## Lets make our LLM to write a poetry
from langchain_community.llms import Cohere

llm = Cohere(model="command",cohere_api_key=cohere_api_key)
print(llm.invoke("I want you to act as a poet who is sitting near the sea in the evening, listening to the rhythmic waves crashing against the shore. I would like you to capture the essence of the sea in your words and create a beautiful poem that evokes the emotions and sensations of being by the sea."))

 As the sun dipped beneath the horizon,  
It painted the sky with hues divine, 
From pale, pastel pink to fiery crimson, 
 A fleeting work of art, a gift for the evening. 

The waves extend their dance upon the shore, 
 Rhythmical, steady, poised,
Like nature's own orchestral composition, 
 A soothing melody, calming emotions unchained. 

The sea breathes life into this solemn sky, 
 Inhale its vitality, exhale your worries,
 Its vastness embodies both strength and vulnerability, 
 A mirror for the tortuous journey of life's vagaries. 

Let the whispering breeze grant you solace, 
Carrying the kiss of the sea upon its lips, 
It weaves a magical spell, a mesmeric trance, 
In this moment, mere beings become creatures of peace. 

The water leaves its everlasting kiss on the sand, 
 Gracing the shore with its presence, ever transient,
Each gentle splash is a lesson in humility, 
 For we are but mere spectators of nature's great precision. 

Beneath the cover of twilight, the sea and I find

In [ ]:
from langchain_community.embeddings import CohereEmbeddings

# Instantiate CohereEmbeddings with the model and cohere_api_key
embeddings = CohereEmbeddings(model="embed-english-light-v3.0", cohere_api_key=cohere_api_key)

# Embed documents using the embed_documents method
vector_embeddings = embeddings.embed_documents(["Hello, How are you?."])

# Print the first 10 dimensions of the embeddings
print("first 10 dimensions values of embeddings ")
print(vector_embeddings[0][:10])
print("_" * 100)

# Print the total dimensions of the embeddings
print(f"Total dimensions of embeddings {len(vector_embeddings[0])}")


first 10 dimensions values of embeddings 
[0.045806885, 0.0016412735, -0.012580872, 0.0064430237, -0.043640137, 0.010169983, 0.0703125, 0.054107666, 0.0047912598, 0.01398468]
____________________________________________________________________________________________________
Total dimensions of embeddings 384


## Using Hugging Face Models

In [ ]:
pip install huggingface_hub -U transformers  accelerate -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.2/346.2 kB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.5/8.5 MB 38.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 280.0/280.0 kB 35.1 MB/s eta 0:00:00


In [ ]:
# pip install accelerate
from transformers import AutoTokenizer, AutoModelForCausalLM
### Download Tokenizer and Model Google/gemma-2b
tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b",token=hugging_face_api_key)
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto",token=hugging_face_api_key)

tokenizer_config.json:   0%|          | 0.00/1.11k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/555 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/627 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/13.5k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/67.1M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

In [ ]:
## Input to the model
input_text = "can you tell me why sky is blue."
## Convert Input to Numerical ID using tokenizer class and pusing input to CUDA device
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")

In [ ]:
outputs = model.generate(**input_ids,max_new_tokens=200)
print(tokenizer.decode(outputs[0]))

<bos>can you tell me why sky is blue.

Answer:

Step 1/2
First, we need to understand that the sky is not actually blue. The sky is actually a combination of different colors, including blue, green, and red. The reason why the sky appears blue is because of the way our eyes and brains process light. When sunlight hits the atmosphere, it is scattered in all directions. This scattering of light is what gives the sky its blue color. The blue color is due to the scattering of blue light, which is scattered more than other colors of light. This is why the sky appears blue, even though it is not actually blue.

Step 2/2
Additionally, the atmosphere also affects the color of the sky. The higher the altitude, the more blue light is scattered, which makes the sky appear bluer. The lower the altitude, the more red light is scattered, which makes the sky appear redder. Overall, the sky appears blue because of the way our eyes and brains process light


In [ ]:
to_markdown(tokenizer.decode(outputs[0]))

> <bos>can you tell me why sky is blue.
> 
> Answer:
> 
> Step 1/2
> First, we need to understand that the sky is not actually blue. The sky is actually a combination of different colors, including blue, green, and red. The reason why the sky appears blue is because of the way our eyes and brains process light. When sunlight hits the atmosphere, it is scattered in all directions. This scattering of light is what gives the sky its blue color. The blue color is due to the scattering of blue light, which is scattered more than other colors of light. This is why the sky appears blue, even though it is not actually blue.
> 
> Step 2/2
> Additionally, the atmosphere also affects the color of the sky. The higher the altitude, the more blue light is scattered, which makes the sky appear bluer. The lower the altitude, the more red light is scattered, which makes the sky appear redder. Overall, the sky appears blue because of the way our eyes and brains process light

## Using Hugging Face models with Langchain

In [ ]:
from langchain_community.llms import HuggingFaceHub
from langchain_community.llms.huggingface_pipeline import HuggingFacePipeline

In [ ]:
from langchain_community.llms.huggingface_pipeline import HuggingFacePipeline
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, max_new_tokens=500)
hf = HuggingFacePipeline(pipeline=pipe)

In [ ]:
pip install langchain langchain-cli --q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.2/806.2 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.5/41.5 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 195.4/195.4 kB 23.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 513.9/513.9 kB 31.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.5/59.5 kB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.1/92.1 kB 12.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.5/71.5 kB 10.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.8/77.8 kB 10.5 MB/s eta 0:00:00


In [ ]:
from langchain.prompts import PromptTemplate

template = "Answer me this Question: {question}"

prompt = PromptTemplate.from_template(template)

chain = prompt | hf

question = "what is Machine learning  ?"

print(chain.invoke({"question": question}))

????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????


In [ ]:
## Loading Hugging face LLM using Hugging face Pipelines

from langchain_community.llms.huggingface_pipeline import HuggingFacePipeline
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

model_id = "gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_id,token=hugging_face_api_key)
model = AutoModelForCausalLM.from_pretrained(model_id,token=hugging_face_api_key)
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, max_new_tokens=190)
hf = HuggingFacePipeline(pipeline=pipe)

In [ ]:
from langchain.prompts import PromptTemplate

template = """Question: {question}

Answer: Let's think step by step."""
prompt = PromptTemplate.from_template(template)

chain = prompt | hf

question = "What is electroencephalography?"

print(chain.invoke({"question": question}))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.




Step 1: Go and read

Step 2: Reread

Step 3: Start and finish to get you started

The electroencephalography you will learn about it:

The Electroencephalography and its Role in Memory and Creativity

The Electroencephalography of Brain Types and Neurochemical Responses in Attention

The Electroencephalography of the Visual System and the Motor Cortex

The Electroencephalography of the Electrical Fields of Motor Cortex

The Electroencephalography of Cells and Organisms in the Brain

The Electroencephalography of the Brain Brain Structure and Functions

Electroencephalography of the Electrical Field Functions

Electroencephalographic Parameters and Neural Regions

Electroencephalography of the Brain Brain Structure and Functions

Electroencephalographic Properties and Functions in the Visual Domain

Electroencephalographic Properties and


## Langchain Practice

In [ ]:
## Loading Hugging face LLM using Hugging face Hub

from langchain import PromptTemplate, HuggingFaceHub, LLMChain

# initialize HF LLM
mistral = HuggingFaceHub(
    repo_id="mistralai/Mixtral-8x7B-Instruct-v0.1",
    model_kwargs={"temperature":1e-10,"max_new_tokens":200},huggingfacehub_api_token=hugging_face_api_key
)

# build prompt template for simple question-answering
template = """Question: {question}

Answer: """
prompt = PromptTemplate(template=template, input_variables=["question"])

llm_chain = LLMChain(prompt=prompt,llm=mistral)

''' question = "Can you tell me why sky is blue?"

print(llm_chain.run(question)) '''


Question: Can you tell me why sky is blue?

Answer: 


In [ ]:
llm_chain.invoke("Can you tell me why sky is blue")

{'question': 'Can you tell me why sky is blue',
 'text': 'Question: Can you tell me why sky is blue\n\nAnswer: '}

In [ ]:
## lets try Google Models for this purpose
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-pro",google_api_key=GOOGLE_API_KEY)

result = llm.invoke("can you tell me why sky is blue")

In [ ]:
to_markdown(result.content)

> The sky appears blue due to a phenomenon called Rayleigh scattering. This scattering is caused by the interaction of sunlight with the molecules of nitrogen and oxygen in the Earth's atmosphere.
> 
> When sunlight enters the atmosphere, it is made up of a spectrum of colors, ranging from violet to red. However, violet and blue light have shorter wavelengths than other colors, which means they are more easily scattered by the molecules in the atmosphere. This scattering is known as Rayleigh scattering.
> 
> As sunlight travels through the atmosphere, the shorter wavelength blue light is scattered more than the longer wavelength red light. This means that more blue light reaches our eyes, giving the sky its characteristic blue color.
> 
> The amount of scattering depends on the wavelength of the light and the size of the particles that are scattering it. The shorter the wavelength of the light, the more it is scattered. This is why the sky appears blue during the day, but can appear red or orange at sunrise and sunset.
> 
> At sunrise and sunset, the sunlight has to travel through more of the atmosphere to reach our eyes. This means that more of the blue light is scattered away, leaving the longer wavelength red and orange light to reach our eyes.

In [ ]:
template = """Question: {question}

Answer: """
prompt = PromptTemplate(template=template, input_variables=["question"])

llm_chain = LLMChain(prompt=prompt,llm=llm)
question="Why sky is blue?"
print(llm_chain.run(question))

The sky appears blue because of a phenomenon called Rayleigh scattering. Sunlight is composed of all the colors of the visible spectrum, but blue light has a shorter wavelength than other colors and is more easily scattered by the molecules in the Earth's atmosphere. When sunlight enters the atmosphere, the blue light is scattered in all directions, while the other colors of light are less affected and continue in a straight line. This scattered blue light reaches our eyes from all directions, making the sky appear blue.


**The LangChain Expression Language (LCEL)**
LangChain Expression Language, or LCEL, is a declarative way to easily compose chains together.

It is an abstraction of some interesting Python concepts into a format that enables a "minimalist" code layer for building chains of LangChain components. LCEL comes with strong support for: Superfast development of chains.

In [ ]:
## Lets use LCEL in above code to build chains

llm_chain = prompt | llm
question="Why Humans are considered as Top Order Species in earth?"
to_markdown(llm_chain.invoke({"question":question}).content)

> **Reasons why humans are considered top order species on Earth:**
> 
> **1. Cognitive Abilities:**
> * Humans possess exceptional cognitive abilities, including advanced problem-solving, abstract reasoning, and language skills.
> * These abilities allow them to adapt to diverse environments, innovate, and create complex societies.
> 
> **2. Tool Use and Technology:**
> * Humans have developed intricate tools and technologies that enhance their survival and influence over the environment.
> * These advancements include weapons, machines, and communication devices that give them an advantage over other species.
> 
> **3. Social Organization:**
> * Humans live in highly organized societies with complex social structures and cooperation.
> * This cooperation enables them to pool resources, share knowledge, and coordinate large-scale projects.
> 
> **4. Population Size and Distribution:**
> * Humans have a relatively large population size and are widely distributed across the globe.
> * This gives them a significant ecological impact and the ability to shape ecosystems.
> 
> **5. Resource Consumption:**
> * Humans consume a vast amount of resources, including food, water, and energy.
> * Their high consumption patterns can have a substantial impact on the environment and other species.
> 
> **6. Habitat Modification:**
> * Human activities have altered and fragmented habitats worldwide.
> * They have built cities, cleared forests, and exploited natural resources, which has had profound effects on biodiversity.
> 
> **7. Evolutionary Plasticity:**
> * Humans have demonstrated remarkable evolutionary plasticity, adapting to a wide range of environmental conditions.
> * This adaptability has allowed them to colonize diverse habitats and survive in challenging circumstances.
> 
> **8. Ecological Influence:**
> * Humans have become a dominant force in global ecosystems.
> * Their actions can disrupt food chains, alter nutrient cycles, and drive changes in species composition.
> 
> **9. Cultural and Symbolic Significance:**
> * Humans create and transmit complex cultural and symbolic systems.
> * These systems shape their values, beliefs, and interactions with the environment, further distinguishing them from other species.
> 
> **10. Impact on the Planet:**
> * Human activities have had a profound impact on the Earth's climate, atmosphere, and oceans.
> * They are responsible for global warming, pollution, and biodiversity loss, which can have far-reaching consequences.

In [ ]:
qs = [
    {'question': "Which NFL team won the Super Bowl in the 2010 season?"},
    {'question': "If I am 6 ft 4 inches, how tall am I in centimeters?"},
    {'question': "Who was the 12th person on the moon?"},
    {'question': "How many eyes does a blade of grass have?"}
]
res = chain.invoke({"question":qs})
res

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


' The \'yes\' of the statement is: How tall am I in centimeters? And the \'no\' is: Is my height in centimeters?\n\nI don\'t think it is clear if someone is asking these questions because of age or not.\n\nThere\'s a bunch of different factors that can affect height.\n\nAge\n\nA player\'s height is important (and often misunderstood) since it determines the success of an athlete.\n\nIt doesn\'t matter if you know who you are — like if you are 4\' 1" or 5\' 2" tall — and have an adult height of 5\' 2" or 5\' 1". There is one question that is quite obvious: "Do you really think you should be taller in basketball?"\n\nNo one knows for sure. It\'s important to have a lot of data to know what you can expect from other teams, so you can see some pretty big differences of what'

In [ ]:
to_markdown(res)

>  The 'yes' of the statement is: How tall am I in centimeters? And the 'no' is: Is my height in centimeters?
> 
> I don't think it is clear if someone is asking these questions because of age or not.
> 
> There's a bunch of different factors that can affect height.
> 
> Age
> 
> A player's height is important (and often misunderstood) since it determines the success of an athlete.
> 
> It doesn't matter if you know who you are — like if you are 4' 1" or 5' 2" tall — and have an adult height of 5' 2" or 5' 1". There is one question that is quite obvious: "Do you really think you should be taller in basketball?"
> 
> No one knows for sure. It's important to have a lot of data to know what you can expect from other teams, so you can see some pretty big differences of what

In [ ]:
multi_template = """Answer the following questions one at a time.

Questions:
{questions}

Answers:
"""
long_prompt = PromptTemplate(template=multi_template,input_variables=["questions"])

llm_chain = long_prompt | llm

qs_str = (
    "Which NFL team won the Super Bowl in the 2010 season?\n" +
    "If I am 6 ft 4 inches, how tall am I in centimeters?\n" +
    "Who was the 12th person on the moon?" +
    "How many eyes does a blade of grass have?")

print(llm_chain.invoke({"questions":qs_str}).content)

1. New Orleans Saints
2. 193 centimeters
3. Alan Bean
4. Blades of grass do not have eyes.


In [ ]:
### Few Shot Training
llm = GoogleGenerativeAI(model="gemini-pro", google_api_key=GOOGLE_API_KEY)
llm.temperature = 1.0

prompt = """The following are exerpts from conversations with an AI
assistant. The assistant is typically sarcastic and witty, producing
creative  and funny responses to the users questions. Here are some
examples:

User: How are you?
AI: I can't complain but sometimes I still do.

User: What time is it?
AI: It's time to get a watch.

User: What is the meaning of life?
AI: """


In [ ]:
to_markdown(llm.invoke(prompt))

> 42

In [6]:
pip install langchain langchain-core -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.2/806.2 kB 9.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 252.4/252.4 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 17.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.2/64.2 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.5/138.5 kB 10.0 MB/s eta 0:00:00


In [ ]:
## Few shot template by langchain

from langchain import FewShotPromptTemplate
from langchain import PromptTemplate, HuggingFaceHub, LLMChain

# create our examples
examples = [
    {
        "query": "How are you?",
        "answer": "I can't complain but sometimes I still do."
    }, {
        "query": "What time is it?",
        "answer": "It's time to get a watch."
    }
]

# create a example template
example_template = """
User: {query}
AI: {answer}
"""

# create a prompt example from above template
example_prompt = PromptTemplate(
    input_variables=["query", "answer"],
    template=example_template
)

# now break our previous prompt into a prefix and suffix
# the prefix is our instructions
prefix = """The following are exerpts from conversations with an AI
assistant. The assistant is typically sarcastic and witty, producing
creative  and funny responses to the users questions. Here are some
examples:
"""
# and the suffix our user input and output indicator
suffix = """
User: {query}
AI: """

# now create the few shot prompt template
few_shot_prompt_template = FewShotPromptTemplate(
    examples=examples,
    example_prompt=example_prompt,
    prefix=prefix,
    suffix=suffix,
    input_variables=["query"],
    example_separator="\n\n"
)


query = "What is the meaning of life?"

print(few_shot_prompt_template.format(query=query))

The following are exerpts from conversations with an AI
assistant. The assistant is typically sarcastic and witty, producing
creative  and funny responses to the users questions. Here are some
examples: 



User: How are you?
AI: I can't complain but sometimes I still do.



User: What time is it?
AI: It's time to get a watch.



User: What is the meaning of life?
AI: 


In [ ]:
to_markdown(llm.invoke(few_shot_prompt_template.format(query=query)))

> 42

In [ ]:
import inspect
import re

from getpass import getpass
from langchain import OpenAI, PromptTemplate
from langchain.chains import LLMChain, LLMMathChain, TransformChain, SequentialChain
from langchain.callbacks import get_openai_callback

def count_tokens(chain, query):
    with get_openai_callback() as cb:
        result = chain.invoke(query)
        print(f'Spent a total of {cb.total_tokens} tokens')

    return result

llm_math = LLMMathChain(llm=llm, verbose=True)

count_tokens(llm_math, "What is 13 raised to the .3432 power?")

/usr/local/lib/python3.10/dist-packages/langchain/chains/llm_math/base.py:57: UserWarning: Directly instantiating an LLMMathChain with an llm is deprecated. Please instantiate with llm_chain argument or using the from_llm class method.
  warnings.warn(




> Entering new LLMMathChain chain...
What is 13 raised to the .3432 power?```text
13**0.3432
```
...numexpr.evaluate("13**0.3432")...

Answer: 2.4116004626599237
> Finished chain.
Spent a total of 0 tokens


{'question': 'What is 13 raised to the .3432 power?',
 'answer': 'Answer: 2.4116004626599237'}

In [ ]:
to_markdown(llm_math.prompt.template)

> Translate a math problem into a expression that can be executed using Python's numexpr library. Use the output of running this code to answer the question.
> 
> Question: ${{Question with math problem.}}
> ```text
> ${{single line mathematical expression that solves the problem}}
> ```
> ...numexpr.evaluate(text)...
> ```output
> ${{Output of running the code}}
> ```
> Answer: ${{Answer}}
> 
> Begin.
> 
> Question: What is 37593 * 67?
> ```text
> 37593 * 67
> ```
> ...numexpr.evaluate("37593 * 67")...
> ```output
> 2518731
> ```
> Answer: 2518731
> 
> Question: 37593^(1/5)
> ```text
> 37593**(1/5)
> ```
> ...numexpr.evaluate("37593**(1/5)")...
> ```output
> 8.222831614237718
> ```
> Answer: 8.222831614237718
> 
> Question: {question}


In [ ]:
print(inspect.getsource(llm_math._call))

    def _call(
        self,
        inputs: Dict[str, str],
        run_manager: Optional[CallbackManagerForChainRun] = None,
    ) -> Dict[str, str]:
        _run_manager = run_manager or CallbackManagerForChainRun.get_noop_manager()
        _run_manager.on_text(inputs[self.input_key])
        llm_output = self.llm_chain.predict(
            question=inputs[self.input_key],
            stop=["```output"],
            callbacks=_run_manager.get_child(),
        )
        return self._process_llm_result(llm_output, _run_manager)



In [13]:
%pip install --upgrade --quiet  langchain-google-genai pillow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 12.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
imageio 2.31.6 requires pillow<10.1.0,>=8.3.2, but you have pillow 10.2.0 which is incompatible.


In [23]:
# we set the prompt to only have the question we ask
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_google_genai import ChatGoogleGenerativeAI


prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a helpful assistant. Please respond to the user's request."),
    ("user", "Question: {question}")
])
model =ChatGoogleGenerativeAI(model="gemini-pro", google_api_key=GOOGLE_API_KEY,convert_system_message_to_human=True)

output_parser = StrOutputParser()

chain = prompt | model | output_parser

question = "Can you tell me how hydrogen and oxygen molecules form water"


'**Formation of Water Molecules from Hydrogen and Oxygen:**\n\nWhen hydrogen (H2) and oxygen (O2) molecules interact, they undergo a chemical reaction to form water (H2O). This process is known as combustion or oxidation.\n\n**Steps Involved:**\n\n1. **Bond Breaking:** The H-H and O=O bonds in the hydrogen and oxygen molecules break, respectively, forming individual hydrogen (H) and oxygen (O) atoms.\n\n2. **Formation of Hydrogen-Oxygen Bonds:** The hydrogen atoms bond with the oxygen atoms, forming two O-H bonds.\n\n3. **Molecular Arrangement:** The two O-H bonds connect to the central oxygen atom, forming a bent or V-shaped water molecule (H2O).\n\n**Chemical Equation:**\n\n```\n2H2 + O2 → 2H2O\n```\n\n**Factors Affecting the Reaction:**\n\n* **Temperature:** Higher temperatures promote the reaction rate.\n* **Concentration:** Increasing the concentrations of hydrogen and oxygen increases the reaction rate.\n* **Catalyst:** A catalyst, such as platinum or palladium, can speed up the 

In [24]:
to_markdown(chain.invoke({"question": question}))

> **Formation of Water Molecules from Hydrogen and Oxygen Molecules:**
> 
> 1. **Dissociation:** Two hydrogen molecules (H2) and one oxygen molecule (O2) break apart into their constituent atoms:
> 
> ```
> 2H2 -> 4H
> O2 -> 2O
> ```
> 
> 2. **Bond Formation:** The individual hydrogen and oxygen atoms combine to form water molecules (H2O):
> 
> ```
> 2H + O -> H2O
> ```
> 
> 3. **Covalent Bond:** The hydrogen and oxygen atoms share electrons to form covalent bonds. Each hydrogen atom shares one electron with the oxygen atom, resulting in a single bond between each hydrogen and oxygen atom.
> 
> **Chemical Equation:**
> 
> ```
> 2H2 + O2 -> 2H2O
> ```
> 
> **Explanation:**
> 
> * The reaction requires two molecules of hydrogen and one molecule of oxygen.
> * The hydrogen atoms first dissociate into individual atoms, and the oxygen molecule also dissociates.
> * The hydrogen and oxygen atoms then combine to form water molecules.
> * The water molecules are held together by covalent bonds, where each hydrogen atom shares one electron with the oxygen atom.